### PR TITLE
Transition-based cascaded parser trainer in C++ using Myelin

### PR DESCRIPTION
**Pull request by ringgaard**
I have implemented the caspar parser trainer in C++ using Myelin. It can train a caspar model based on OntoNotes in roughly 90 minutes (~5000 tokens/sec), achieving accuracy that is similar to the PyTorch-based parser (81% SLOT F1).

I have factored out the parser feature extractor from the parser runtime so it can be shared with the parser trainer. The parser state has also be refactored to include all the state needed for feature extraction.

The parser trainer is implemented as a base class that can be specialized for different setups. The caspar-trainer is one specialization that uses the same two-cascade setup as the PyTorch-based trainer.

The following components have been deprecated/replaced:
* sling/base/strtoint
* sling/nlp/document/document-source (in favor of document-corpus)
* sling/util/table-writer

### CODE REVIEW

File: sling/nlp/parser/tools/train_caspar.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+trainer = wf.task("caspar-trainer")
+
+trainer.add_params({
+  "learning_rate": 1.0,
+  "learning_rate_decay": 0.8,
+  "clipping": 1,
+  "optimizer": "sgd",
+  "epochs": 50000,
+  "batch_size": 32,
+  "learning_rate": 1.0,
```
Comment (anders-sandholm) [inline sling/nlp/parser/tools/train_caspar.py]: you can safely delete this line (already appears above)

File: sling/nlp/parser/caspar-trainer.cc [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      for (const Token &t : document->tokens()) words_[t.word()]++;
+
+      // Generate action table for fine-grained classifier.
+      Generate(*document, [&](const ParserAction &action) {
+        bool skip = false;
+        switch (action.type) {
+          case ParserAction::SHIFT:
+          case ParserAction::MARK:
+            skip = true;
+            return;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/parser/caspar-trainer.cc]: Good point. Done.

File: sling/nlp/parser/parser-state.cc [hunk 3]
```diff
@@ -88,25 +88,18 @@ void ParserState::Apply(const ParserAction &action) {
...
-  return false;
+  step_++;
 }
 
 bool ParserState::CanApply(const ParserAction &action) const {
   if (done_) return false;
   switch (action.type) {
     case ParserAction::CASCADE:
-      return false;
+      // Do not allow cacading back to the main cascade.
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/nlp/parser/parser-state.cc]: Fixed.

File: sling/nlp/parser/tools/train_caspar.py (hunk 1 above)
Comment (ringgaard) [reply sling/nlp/parser/tools/train_caspar.py]: Oops. Fixed.
